overlord/configstate/configcore/tmp.go

```diff
@@ -0,0 +1,147 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/gadget/quantity"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/sysconfig"
+)
+
+const (
+	mntStaticOptions         = "mode=1777,strictatime,nosuid,nodev"
+	tmpfsMountPoint          = "/tmp"
+	tmpMntServOverrideSubDir = "tmp.mount.d"
+	tmpMntServOverrideFile   = "override.conf"
+)
+
+func init() {
+	// add supported configuration of this module
+	supportedConfigurations["core.tmp.size"] = true
+}
+
+func validTmpfsSize(sizeStr string) error {
+	if sizeStr == "" {
+		return nil
+	}
+
+	// TODO allow also percentages. That is allowed for CPU quotas so
+	// it is probably fine to allow that for tmp.size too.
+	size, err := quantity.ParseSize(sizeStr)
+	if err != nil {
+		return err
+	}
+
+	// Do not allow less than 16mb
+	// 0 is special and means unlimited
+	if size > 0 && size < 16*quantity.SizeMiB {
+		return fmt.Errorf("size is less than 16Mb")
+	}
+
+	return nil
+}
+
+func validateTmpfsSettings(tr config.ConfGetter) error {
+	tmpfsSz, err := coreCfg(tr, "tmp.size")
+	if err != nil {
+		return err
+	}
+
+	return validTmpfsSize(tmpfsSz)
+}
+
+func handleTmpfsConfiguration(_ sysconfig.Device, tr config.ConfGetter, opts *fsOnlyContext) error {
+	tmpfsSz, err := coreCfg(tr, "tmp.size")
+	if err != nil {
+		return err
+	}
+
+	// Create override configuration file for tmp.mount service
+
+	// Create /etc/systemd/system/tmp.mount.d if needed
+	var overrDir string
+	if opts == nil {
+		// runtime system
+		overrDir = dirs.SnapServicesDir
+	} else {
+		overrDir = dirs.SnapServicesDirUnder(opts.RootDir)
+	}
+	overrDir = filepath.Join(overrDir, tmpMntServOverrideSubDir)
+
+	// Write service config override if needed
+	options := mntStaticOptions
+	dirContent := make(map[string]osutil.FileState, 1)
+	cfgFilePath := filepath.Join(overrDir, tmpMntServOverrideFile)
+	modify := true
+	if tmpfsSz != "" {
+		if err := os.MkdirAll(overrDir, 0755); err != nil {
+			return err
+		}
+		options = fmt.Sprintf("%s,size=%s", options, tmpfsSz)
+		content := fmt.Sprintf("[Mount]\nOptions=%s\n", options)
+		dirContent[tmpMntServOverrideFile] = &osutil.MemoryFileState{
+			Content: []byte(content),
+			Mode:    0644,
+		}
+		oldContent, err := ioutil.ReadFile(cfgFilePath)
```

Comment:
Do we need a `if err != nil ` here or is the error handled elsewhere?

Reply:
Good question... here we want to check the content of the file just to know if we need to remount or not. If there was an error it will be usually because `override.conf` does not exist, thing that will happen when setting `tmp.size` for the first time and we don't care too much about that as we will write the file later when calling `osutil.EnsureDirState()`. The question that could be raised is what to do if the error is something different like permissions error, for instance. And for that case things will fail probably later in the call to  `osutil.EnsureDirState()`. Maybe we could fail before that, but I have the feeling that it is preferable to fail in the Ensure call, which might be more robust and implement more cases.